File: src/day-8/scala3/src/main/scala/Part1.scala
case class Node(baseNode: String, routes: List[String])

def data: List[String] = {
  import scala.io.Source
  val plik = Source.fromFile("src/main/data/data.txt", "UTF-8")
  plik.getLines.toList
}

def getDirections(dataString: List[String]): List[Node] = {
  val directionsArray = dataString.slice(2, dataString.length).map((row) => row.split(" = ").toList)

  val directions = directionsArray.foldLeft(List[Node]())((directionsList, directionRow) => {
    val baseNode = directionRow(0)
    val routes = directionRow(1).replaceAll("[()]", "").split(", ").toList
    directionsList :+ Node(baseNode, routes)
  })
  directions
}

def calculateSteps(instruction: String, directions: List[Node]) = {
  val directionValues: Map[String, Int] = Map(
    "L" -> 0,
    "R" -> 1
  )

  val startNode: Node = directions.find((node: Node) => node.baseNode == "AAA").get
  val translatedInstructions: List[Int] =
    instruction.split("").toList.map((letter: String) => directionValues(letter))

  def traverse(currInstructions: List[Int], currNode: Node, steps: Int = 0): Int = {
    if (currNode.baseNode == "ZZZ") steps
    else if (currInstructions.length == 0) traverse(translatedInstructions, currNode, steps)
    else
      traverse(
        currInstructions.tail,
        directions.find((node: Node) => node.baseNode == currNode.routes(currInstructions(0))).get,
        steps + 1
      )
  }

  traverse(translatedInstructions, startNode)
}

@main def part1: Unit = {
  val instructions: String = data(0)
  val directions: List[Node] = getDirections(data)
  val steps: Int = calculateSteps(instructions, directions)
  println(steps)
}

File: src/day-8/typescript/part1.ts
// import { testData12 as data } from './data/part1-test-data';
import { data1 as data } from './data/part1-data';

interface DirectionValue {
  [key: string]: number;
}

interface Nodes {
  [key: string]: string[];
}

const directionValues: DirectionValue = {
  'L': 0,
  'R': 1
};

const getDirections = (dataString: string): Nodes => {
  const directionsArray: string[][] = dataString
    .split('\n')
    .slice(2)
    .map((row: string) => row.split(' = '));

  const directions: Nodes = directionsArray.reduce((directionsList: Nodes, directionRow: string[]) => {
    const baseNode: string = directionRow[0];
    const routes: string[] = directionRow[1].replace(/[()]/g, '').split(', ');
    return { ...directionsList, [baseNode]: routes };
  }, {});

  return directions;
};

function calculateSteps(instruction: string, directions: Nodes): number {
  function traverse(currInstruction: number[], currNode: string, steps: number = 0): number {
    if (currNode === 'ZZZ') return steps;
    if (currInstruction.length === 0) currInstruction = translatedInstructions;
    // console.log(currNode, steps);

    return traverse(currInstruction.slice(1), directions[currNode][currInstruction[0]], steps + 1);
  }

  const translatedInstructions: number[] = instruction.split('').map((letter: string) => directionValues[letter]);

  return traverse(translatedInstructions, 'AAA');
}

const intruction: string = data.split('\n')[0];
const directions: Nodes = getDirections(data);
// console.log(intruction);
// console.log(directions);

const steps: number = calculateSteps(intruction, directions);
console.log(steps);
